File: f1tenth_waypoint/src/utils/waypoint_utils.cpp
#include "f1tenth_waypoint/waypoint_utils.hpp"


CSVFile::CSVFile(std::shared_ptr<rclcpp::Node> rclcpp_node) : node_(rclcpp_node)
{
    this->initialize_parameters();

    /*std::filesystem::current_path()  return /home/fablab_01*/
    path_to_csvFile = std::filesystem::current_path() / this->csv_path / csvFile_name;

    RCLCPP_INFO(this->node_->get_logger(), "full_path %s", path_to_csvFile.c_str());

    this->initialize_csv(path_to_csvFile);


}


void CSVFile::initialize_parameters()
{
    try
    {

        // Check and declare parameters if they don't exist
        if (!node_->has_parameter("path_to_csvFile")) {
            this->node_->declare_parameter<std::string>("path_to_csvFile", "/home/fablab_01/f1_ws/src/f1tenth_projects/f1tenth_waypoint/racelines");
        }

        if (!node_->has_parameter("csvFile_name")) {
            this->node_->declare_parameter<std::string>("csvFile_name", "f1tenth_waypoint_path.csv");
        }
        // Retrieve parameters

        csv_path = this->node_->get_parameter("path_to_csvFile").as_string();
        csvFile_name = this->node_->get_parameter("csvFile_name").as_string();

        RCLCPP_INFO(this->node_ ->get_logger(), "CSV Path: %s", csv_path.c_str());
        RCLCPP_INFO(this->node_ ->get_logger(), "csvfile name %s", csvFile_name.c_str());

        }
        catch(const rclcpp::exceptions::ParameterNotDeclaredException &e) 
        {
            RCLCPP_WARN(this->node_ -> get_logger(), "Could not get parameters ... setting variables to default");
            RCLCPP_WARN(this->node_ -> get_logger(), "Error: %s", e.what());
        }
        catch(const std::exception &e)
        {
            RCLCPP_WARN(this->node_ ->get_logger(), "An error occured: %s", e.what());
            
        }

        

}

void CSVFile::initialize_csv(std::string path_to_csv_)
{
    // ofstream FileOut;
    // std::string header_name;

    std::ofstream csv_file(path_to_csv_, std::ios::out | std::ios::trunc);
    if(!csv_file.is_open())
    {
        std::cerr << "Failed to create or open CSV file " << path_to_csv_ << std::endl;
        return;
    }
    RCLCPP_INFO(this->node_ ->get_logger(), "Creating or overwriting CSV file at: %s", path_to_csv_);
    csv_file << "x, y, yaw" << std::endl;
    csv_file.close();
}


void CSVFile::write_points_to_csv(const geometry_msgs::msg::Point::ConstSharedPtr waypoint_msg_)
{
    std::ofstream csv_file(path_to_csvFile, std::ios::app);
    if(!csv_file.is_open())
    {
        std::cerr << "Failed to open CSV file: " << path_to_csvFile << std::endl;
        return;
    }
    double x = waypoint_msg_->x;
    double y = waypoint_msg_->y;
    double yaw = waypoint_msg_->z;
    std::cout << "Writing Clicked Point: x=" << x << ", y=" << y << ", yaw=" << yaw << std::endl;
    csv_file << x << "," << y << "," << yaw << std::endl;
    csv_file.close();


}
void CSVFile::write_points_to_csv(const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg_)
{
    std::ofstream csv_file(path_to_csvFile, std::ios::app);
    if(!csv_file.is_open())
    {
        std::cerr << "Failed to open CSV file: " << path_to_csvFile << std::endl;
        return;
    }

    double x = odom_msg_->pose.pose.position.x;
    double y = odom_msg_->pose.pose.position.y;

    // Extract quaternion and convert to yaw

    const geometry_msgs::msg::Quaternion &quat_msg = odom_msg_->pose.pose.orientation;

    tf2::Quaternion quat;
    tf2::fromMsg(quat_msg, quat); tf2::Matrix3x3 mat(quat);
    
    double roll, pitch, yaw;
    mat.getRPY(roll, pitch, yaw);



    std::cout << "Writing Odometry: x=" << x << ", y=" << y << ", yaw=" << yaw << std::endl;

    // Write row to CSV
    csv_file << x << "," << y << "," << yaw << std::endl;
    csv_file.close();


}


Waypoint_Visualization::Waypoint_Visualization(std::shared_ptr<rclcpp::Node> rclcpp_node) : node_(rclcpp_node)
{
    this->initialize_marker();
}

Waypoint_Visualization::~Waypoint_Visualization()
{

}


void Waypoint_Visualization::initialize_marker()
{
    this->waypoint_marker_.header.frame_id = "map";
    this->waypoint_marker_.type = visualization_msgs::msg::Marker::LINE_STRIP;
    this->waypoint_marker_.action = visualization_msgs::msg::Marker::ADD;

    this->waypoint_marker_.scale.x = 0.05;
    this->waypoint_marker_.scale.y = 0.1;
    this->waypoint_marker_.scale.z = 0.1;

    this->waypoint_marker_.color.a = 1.0;
    this->waypoint_marker_.color.r = 0.0;
    this->waypoint_marker_.color.g = 1.0;
    this->waypoint_marker_.color.b = 0.0;
}

void Waypoint_Visualization::update_waypoint_marker_callback(const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg_)
{
    geometry_msgs::msg::Point point;
    point.x = odom_msg_->pose.pose.position.x;
    point.y = odom_msg_->pose.pose.position.y;
    point.z = 0.0;

    this->waypoint_marker_.points.push_back(point);

}

 void Waypoint_Visualization::update_waypoint_marker_callback(const geometry_msgs::msg::Point::ConstSharedPtr waypoint_msg_)
 {
    this->waypoint_marker_.points.push_back(*waypoint_msg_);
}


visualization_msgs::msg::Marker &Waypoint_Visualization::get_waypoint_marker()
{
    return waypoint_marker_;
}



/*===================================================================*/

UtilsWaypoint::UtilsWaypoint(std::shared_ptr<rclcpp::Node> rclcpp_node) : node_(rclcpp_node)
{
   csv_file = std::make_shared<CSVFile>(rclcpp_node);
   waypoint_visualization = std::make_shared<Waypoint_Visualization>(rclcpp_node);
}


void UtilsWaypoint::write_waypoint_to_csv(const geometry_msgs::msg::Point::ConstSharedPtr waypoint_msg_)
{
   this->csv_file -> write_points_to_csv(waypoint_msg_);
}
void UtilsWaypoint::write_waypoint_to_csv(const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg_)
{
    this->csv_file -> write_points_to_csv(odom_msg_);
}


void UtilsWaypoint::update_waypoint_marker_callback(const geometry_msgs::msg::Point::ConstSharedPtr waypoint_msg_)
{
    this->waypoint_visualization -> update_waypoint_marker_callback(waypoint_msg_);
}
void UtilsWaypoint::update_waypoint_marker_callback(const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg_)
{
    this->waypoint_visualization -> update_waypoint_marker_callback(odom_msg_);
}


// visualization_msgs::msg::Marker UtilsWaypoint::get_waypoint_marker() const
// {
//     return this->waypoint_visualization;
// }


visualization_msgs::msg::Marker &UtilsWaypoint::get_waypoint_marker()
{
    // Delegate the call to the instance of Waypoint_Visualization
    return waypoint_visualization->get_waypoint_marker();
}


UtilsWaypoint::~UtilsWaypoint()
{

}

File: f1tenth_waypoint/f1tenth_waypoint_generator/utils/waypoint_utils.py
import csv
import os
from typing import List, Tuple, Union

from rclpy.node import Node

from visualization_msgs.msg import Marker
from geometry_msgs.msg import PointStamped, Point
from nav_msgs.msg import Odometry
from tf_transformations import euler_from_quaternion

class CSVFile:
    def __init__(self, rclpy_node: Node):
        self.rclpy_node = rclpy_node

        # rclpy_node.get_logger().info('Initializing parameters')

        if not self.rclpy_node.has_parameter('path_to_csvFile'):
            rclpy_node.declare_parameter(name='path_to_csvFile', value="/f1_ws/src/f1tenth_system/f1tenth_waypoint/racelines")
        
        if not self.rclpy_node.has_parameter('csvFile_name'):
            rclpy_node.declare_parameter(name='csvFile_name', value="f1tenth_waypoint.csv")

        try:
            csv_path = rclpy_node.get_parameter('path_to_csvFile').get_parameter_value().string_value
            csvFile_name = rclpy_node.get_parameter('csvFile_name').get_parameter_value().string_value

        except Exception as e:
            self.rclpy_node.get_logger().warn('Could not get parameters ...setting variables to default')
            self.rclpy_node.get_logger().warn(f'Error: "{e}"')

        # Construct the full path to the CSV file
        self.path_to_csvFile = os.path.join(os.getcwd() , csv_path, csvFile_name )

        # Initialize the CSV file by writing a header
        self.initialize_csv(self.path_to_csvFile )
    
    def initialize_csv(self, csv_path:str):

        """ Initialize the CSV file and write the header. Overwrites the file. """

        with open(csv_path, mode='w', newline='') as csv_file:
            csv_writer = csv.writer(csv_file, delimiter=',')
            self.rclpy_node.get_logger().info(f"Creating or overwriting CSV file at {csv_path} and writing header.")
            csv_writer.writerow(['x', 'y', 'heading'])
    
    def write_points_to_csv(self, waypoint_msgs:Union[Point, Odometry]) -> None:

        """ Write points from PointStamped or Odometry message to the CSV file. """

        with open(self.path_to_csvFile, mode='a', newline='') as csv_file:
            csv_writer = csv.writer(csv_file, delimiter=',')
    
            if isinstance(waypoint_msgs, Point):
                x = waypoint_msgs.x 
                y = waypoint_msgs.y  
                yaw = waypoint_msgs.z 
                self.rclpy_node.get_logger().info(f'Writing Clicked Point: x={x}, y={y}, yaw={yaw}')
                csv_writer.writerow([x, y, yaw])
                
            elif isinstance(waypoint_msgs, Odometry):
                # Extract x, y from Odometry
                x = waypoint_msgs.pose.pose.position.x
                y = waypoint_msgs.pose.pose.position.y
                quaternion = (waypoint_msgs.pose.pose.orientation.x, waypoint_msgs.pose.pose.orientation.y, 
                              waypoint_msgs.pose.pose.orientation.z, waypoint_msgs.pose.pose.orientation.w)

                _,_,yaw = euler_from_quaternion(quaternion)
                self.rclpy_node.get_logger().info(f'Writing Odometry: x={x}, y={y}, yaw={yaw}')
                csv_writer.writerow([x, y, yaw])

            else:
                self.rclpy_node.get_logger().warn(f"Unknown message type received: {type(waypoint_msgs)}")

class Waypoint_Visualization:
    def __init__(self, rclpy_node: Node) -> None:
        self.rclpy_node = rclpy_node
        self.waypoint_marker = Marker()
        self.initialize_marker()
    
    def initialize_marker(self)->None:
        self.waypoint_marker.header.frame_id = "map"
        self.waypoint_marker.type = Marker.LINE_STRIP
        self.waypoint_marker.action = Marker.ADD
        self.waypoint_marker.scale.x = 0.05 # width of the line 
        self.waypoint_marker.scale.y = 0.1 
        self.waypoint_marker.scale.z = 0.1 

        self.waypoint_marker.color.a = 1.0 
        self.waypoint_marker.color.r = 0.0
        self.waypoint_marker.color.g = 1.0
        self.waypoint_marker.color.b = 0.0

    def update_waypoint_marker_callback(self, waypoint_msgs:Union[Point, Odometry])->None:
        """ Updates the marker with new waypoints."""
        if isinstance(waypoint_msgs, Point):
            self.waypoint_marker.points.append(waypoint_msgs)

        elif isinstance(waypoint_msgs, Odometry):
            point = Point()
            point.x = waypoint_msgs.pose.pose.position.x 
            point.y = waypoint_msgs.pose.pose.position.y
            point.z = 0.0
            self.waypoint_marker.points.append(point)

        else:
            # self.rclpy_node.get_logger().warn(f"Unknown message type received: {type(waypoint_msgs)}")
            raise ValueError(f"Unknown message type received: {type(waypoint_msgs)}")
    
    def get_waypoint_marker(self)->Marker:
        """Returns the current marker object for publishing."""
        return self.waypoint_marker




class UtilsWaypoint():
    def __init__(self, rclpy_node: Node):
        """ Initialize the utility class with the Node and a CSVFile instance """

        self.rclpy_node = rclpy_node

        self.csv_file = CSVFile(self.rclpy_node)

        self.waypoint_marker = Waypoint_Visualization(self.rclpy_node)

    def process_waypoint(self, waypoint_msgs: Union[Point, Odometry]) -> None:
        """ Process the waypoint and write to CSV """
        self.csv_file.write_points_to_csv(waypoint_msgs)
    
    def update_waypoint_marker(self, waypoint_msgs: Union[Point, Odometry]) -> None:
        self.waypoint_marker.update_waypoint_marker_callback(waypoint_msgs)
    
    def get_waypoint_marker(self) -> Marker:
        """Return the current waypoint marker"""
        return self.waypoint_marker.get_waypoint_marker()


